File: frontend/src/service/UserStatsService.ts
import { ApiService } from '@/service/ApiService.ts'
import type { ApiResponse } from '@/DTO/response/ApiResponse.ts'
import type { Page } from '@/DTO/response/Page.ts'
import type { UserStatsResponse } from '@/DTO/response/UserStatsResponse.ts'

export class UserStatsService {
  private ApiService: ApiService

  constructor() {
    this.ApiService = new ApiService()
  }

  public async getLeaderboard(
    page: number,
    size: number,
    sortBy: string,
    direction: string,
  ): Promise<ApiResponse<Page<UserStatsResponse>>> {
    return await this.ApiService.get(
      `user-stats/public/leaderboard?page=${page}&size=${size}&sortBy=${sortBy}&direction=${direction}`,
    )
  }
}

File: Assets/Scripts/Service/UserStatsService.cs
using System.Collections;
using System.Collections.Generic;
using DTO.Response;
using DTO.Response.Statistics;
using DTO.Response.User;
using UnityEngine;
using Utils;

namespace Service
{
    public class UserStatsService
    {
        public IEnumerator GetTop20HighestStreaks(
            System.Action<ApiResponse<List<HighestStreakRankingDTO>>> onSuccess,
            System.Action<ApiErrorResponse<Void>> onError)
        {
            yield return ApiClient.Get<ApiResponse<List<HighestStreakRankingDTO>>, ApiErrorResponse<Void>>(
                "/user-stats/public/top-20-highest-streaks",
                response => onSuccess?.Invoke(response),
                error => onError?.Invoke(error)
            );
        }

        public IEnumerator GetTop20CurrentStreaks(
            System.Action<ApiResponse<List<CurrentStreakRankingDTO>>> onSuccess,
            System.Action<ApiErrorResponse<Void>> onError)
        {
            yield return ApiClient.Get<ApiResponse<List<CurrentStreakRankingDTO>>, ApiErrorResponse<Void>>(
                "/user-stats/public/top-20-current-streaks",
                response => onSuccess?.Invoke(response),
                error => onError?.Invoke(error)
            );
        }

        public IEnumerator GetUserStreak(
            System.Action<ApiResponse<UserStreakDTO>> onSuccess,
            System.Action<ApiErrorResponse<Void>> onError)
        {

            if (User.IsLoggedIn())
            {
                yield return ApiClient.Get<ApiResponse<UserStreakDTO>, ApiErrorResponse<Void>>(
                    "/user-stats/get-streak",
                    response => onSuccess?.Invoke(response),
                    error => onError?.Invoke(error),
                    User.GetToken()
                );
            }
        }
    }
}